### PR TITLE
dmd: Move all public C++ functions into a dmd namespace

### PR DESCRIPTION
**Pull request by ibuclaw**
As discussed @kinke / @RazvanN7.

`extern(C++)` can/should be move to separate PR - I've just bundled it all in here, as it was simpler to fix them up so they didn't appear in the regexp I was using to match functions as I was going through them.

---

Update: now all `extern(C++)` top-level functions are in one (boilerplate) leaf module.

### CODE REVIEW

File: compiler/src/dmd/frontend.h [hunk 1]
```diff
@@ -7805,16 +7976,91 @@ struct Target final
...
+}
+namespace dmd
+{
+}
+namespace dmd
+{
+}
+namespace dmd
+{
+}
```
Comment (ibuclaw) [inline compiler/src/dmd/frontend.h:8043]: I have no idea what's going on here with dtoh. It's harmless, but likely warrants a bug report.

**Comment by ibuclaw** [comment]
Might have to park this for a while, the ability to import symbols in C++ "namespaces" wasn't added until 2.087.

https://issues.dlang.org/show_bug.cgi?id=19893

Workaround is to (1) remove all selective imports, and (2) add these in the same module as the definition.
```
static if (__VERSION__ < 2087)
{
    public Expression ctfeInterpret(Expression e)
    {
        return __traits(getMember, dmd.dinterpret.dmd, "ctfeInterpret")(e);
    }
}
```

**Comment by ibuclaw** [comment]
> Might have to park this for a while, the ability to import symbols in C++ "namespaces" wasn't added until 2.087.
> 
> https://issues.dlang.org/show_bug.cgi?id=19893
> 
> Workaround is to (1) remove all selective imports, and (2) add these in the same module as the definition.
> ```
> static if (__VERSION__ < 2087)
> {
>     public Expression ctfeInterpret(Expression e)
>     {
>         return __traits(getMember, dmd.dinterpret.dmd, "ctfeInterpret")(e);
>     }
> }
> ```


Sitting on this, I think alternatively this could be turned on its head.

1. Remove `extern(C++)` from all these functions.

```
// dmd.mtype
extern (D) Type mutableOf(Type t) { ... }
```

2. Create a new leaf module for exposing D code to the C++ API.
```
module dmd.cxxfrontend;
static import dmd.mtype;
extern(C++, "dmd"):
Type mutableOf(Type t)
{
    return dmd.mtype.mutableOf(t);
}
```
@kinke any better ideas?

**Comment by ibuclaw** [comment]
Now all C++ definitions have been put in a leaf module - not compiled into DMD, but for the C++ FE unittester only - there's nothing blocking this. We can continue having 2.079 as our baseline version, this module is only for GDC and LDC anyway.

**Comment by ibuclaw** [comment]
@kinke I'm yet to give this a test against gdc, but please do have a glance at it regardless.  I've put a few functions in `version(IN_LLVM)` because gdc doesn't call them from the C++ side.

**Comment by ibuclaw** [comment]
One `extern(C++)` definition escaped my regex -  #16182 - with that additional change, everything links.  For my own sanity, have tested bootstrapping from gdc-9, gdc-11, and gdc-13.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Looks good to me. Mind the conflict and we are good to go.

@kinke any objections?

**Comment by kinke** [comment]
No objections, should be good. I probably won't be able to check this before merging v2.108 into LDC, so might follow up then.

**Pull request merged**